contracts/CollateralRegistry.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: bsl-1.1
+
+/*
+  Copyright 2020 Unit Protocol: Artem Zakharov (az@unit.xyz).
+*/
+pragma solidity ^0.7.1;
+pragma experimental ABIEncoderV2;
+
+
+import "./VaultParameters.sol";
+
+
+/**
+ * @title CollateralRegistry
+ **/
+contract CollateralRegistry is Auth {
+
+    event CollateralAdded(address indexed asset);
+    event CollateralRemoved(address indexed asset);
+
+    mapping(address => uint) public collateralId;
+
+    address[] _collaterals;
+    
+    constructor(address _vaultParameters) Auth(_vaultParameters) {}
+
+    function addCollateral(address asset) public onlyManager {
+        require(asset != address(0), "Unit Protocol: ZERO_ADDRESS");
+
+        if (_collaterals.length != 0) {
+            uint id = collateralId[asset];
+            require(id == 0 && _collaterals[0] != asset, "Unit Protocol: ALREADY_EXIST");
+        }
+
+        collateralId[asset] = _collaterals.length;
+        _collaterals.push(asset);
+
+        emit CollateralAdded(asset);
+    }
+
+    function removeCollateral(address asset) public onlyManager {
+        require(asset != address(0), "Unit Protocol: ZERO_ADDRESS");
+
+        uint id = collateralId[asset];
+        require(id != 0 || _collaterals[0] == asset, "Unit Protocol: DOES_NOT_EXIST");
+
+        delete collateralId[asset];
+
+        uint lastId = _collaterals.length - 1;
+        address lastCollateral = _collaterals[lastId];
+        _collaterals[id] = lastCollateral;
+        collateralId[lastCollateral] = id;
```

Comment:
Incorrect if `lastId == id`.

Reply:
Tests should be written covering special cases: deletion of the first and last elements.

Reply:
Added test cases, works just fine with this logic

Reply:
`expect(await this.collateralRegistry.collateralId(collateralAddress3)).to.be.bignumber.equal(new BN(0));`
fails after the removal